### ISSUE TITLE
Type checking for higher order function with variadic generics fails without variable assignment

### ISSUE THREAD

**Issue by sylee957** [microsoft/pyright#5113]
**Describe the bug**

Given the functions:

```python
from __future__ import annotations

from collections.abc import Callable
from typing import TypeVar, TypeVarTuple

X = TypeVar("X")
Y = TypeVar("Y")
Xs = TypeVarTuple("Xs")
Ys = TypeVarTuple("Ys")


def nil() -> tuple[()]:
    return ()


def cons(
    f: Callable[[X], Y],
    g: Callable[[*Xs], tuple[*Ys]],
) -> Callable[[X, *Xs], tuple[Y, *Ys]]:
    def wrapped(x: X, *xs: *Xs) -> tuple[Y, *Ys]:
        y, ys = f(x), g(*xs)
        return y, *ys

    return wrapped


def star(f: Callable[[X], Y]) -> Callable[[*tuple[X, ...]], tuple[Y, ...]]:
    def wrapped(*xs: X):
        if not xs:
            return nil()
        return cons(f, star(f))(*xs)

    return wrapped
```

I get an error

```
test.py:31:29 - error: Argument of type "(X@star) -> Y@star" cannot be assigned to parameter "f" of type "(X@star) -> Y@star" in function "star"
    Type "(X@star) -> Y@star" cannot be assigned to type "(X@star) -> Union[*Ys@cons]"
      Function return type "Y@star" is incompatible with type "Union[*Ys@cons]"
        Type "Y@star" cannot be assigned to type "Union[*Ys@cons]" (reportGeneralTypeIssues)
1 error, 0 warnings, 0 informations
```

However, if I use the walrus trick to store the temporary variable, the issue is not present.

```python
def star(f: Callable[[X], Y]) -> Callable[[*tuple[X, ...]], tuple[Y, ...]]:
    def wrapped(*xs: X):
        if not xs:
            return nil()
        return cons(f, _ := star(f))(*xs)

    return wrapped

```

The issue is also not present in pyright `1.1.305`, and appears after update.

**VS Code extension or command-line**

I'm using pyright `1.1.308`

**Additional context**

**Comment by erictraut** [comment]
Thanks for the bug report. This will be addressed in the next release of pyright.

**Comment by erictraut** [comment]
This is included in pyright 1.1.309, which I just published. It will also be included in a future release of pylance.
